From 19a4b476e04959f2bea8baa482011d07769617ce Mon Sep 17 00:00:00 2001
From: Jonas Gorski <jonas.gorski@bisdn.de>
Date: Wed, 27 May 2026 13:35:12 +0200
Subject: [PATCH 1/2] delta-ag5648: convert to I2C mux and use optoe

Instead of implementing its own SFP EEPROM handling, convert the sfp
driver to register itself as an I2C mux with one bus per SFP port, and
then use the optoe driver for the ports.

This significantly simplifies the driver, and aligns the code with other
platforms.

Since the mux will block the address in all muxed busses, we need to
move it away from the default 0x50 / 0xA0 address, else it conflicts
with the optoe driver. The actual address does not matter, as it does
not read from the muxed bus / address, but directly accesses the master
CPLD on a different bus. So use 0x31, which matches the address of the
master CPLD.

Upstream-Status: Inactive-Upstream [lastcommit: 2024-09-21]

Signed-off-by: Jonas Gorski <jonas.gorski@bisdn.de>
---
 .../ag5648/modules/builds/dni_ag5648_sfp.c    | 255 +++---------------
 .../module/src/platform_lib.h                 |  12 +-
 .../x86_64_delta_ag5648/module/src/sfpi.c     |  22 +-
 3 files changed, 52 insertions(+), 237 deletions(-)

diff --git a/packages/platforms/delta/x86-64/ag5648/modules/builds/dni_ag5648_sfp.c b/packages/platforms/delta/x86-64/ag5648/modules/builds/dni_ag5648_sfp.c
index 849f4b7044c0..8b51779178ad 100755
--- a/packages/platforms/delta/x86-64/ag5648/modules/builds/dni_ag5648_sfp.c
+++ b/packages/platforms/delta/x86-64/ag5648/modules/builds/dni_ag5648_sfp.c
@@ -25,6 +25,7 @@
 #include <linux/module.h>
 #include <linux/init.h>
 #include <linux/i2c.h>
+#include <linux/i2c-mux.h>
 #include <linux/slab.h>
 #include <linux/mutex.h>
 #include <linux/device.h>
@@ -69,32 +70,16 @@
 
 long sfp_port_data = 0;
 
-static const u8 cpld_to_port_table[] = {
-	0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 
-  0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f, 
-  0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 
-  0x18, 0x19, 0x1a, 0x1b, 0x1c, 0x1d, 0x1e, 0x1f, 
-  0x20, 0x21, 0x22, 0x23, 0x24, 0x25, 0x26, 0x27, 
-  0x28, 0x29, 0x2a, 0x2b, 0x2c, 0x2d, 0x2e, 0x2f, 
-  0x30, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36 };
-
 /* Addresses scanned */
-static const unsigned short normal_i2c[] = { 0x50, I2C_CLIENT_END };
+static const unsigned short normal_i2c[] = { 0x31, I2C_CLIENT_END };
 
 /* Each client has this additional data */
 struct ag5648_sfp_data 
 {
   struct device	*hwmon_dev;
-  struct mutex update_lock;
-  char valid;
-  unsigned long last_updated;
   int port;
-  char eeprom[256];
 };
 
-static ssize_t for_eeprom(struct device *dev, struct device_attribute *dev_attr,char *buf);
-static int ag5648_sfp_read_block(struct i2c_client *client, u8 command,u8 *data, int data_len);
-static struct ag5648_sfp_data *ag5648_sfp_update_device(struct device *dev);
 static ssize_t for_status(struct device *dev, struct device_attribute *dev_attr, char *buf);
 static ssize_t set_w_port_data(struct device *dev, struct device_attribute *dev_attr, const char *buf, size_t count);
 static ssize_t for_r_port_data(struct device *dev, struct device_attribute *dev_attr, char *buf);
@@ -106,7 +91,6 @@ extern int i2c_cpld_write(int bus, unsigned short cpld_addr, u8 reg, u8 value);
 
 enum ag5648_sfp_sysfs_attributes 
 {
-  SFP_EEPROM,
   SFP_SELECT_PORT,
   SFP_IS_PRESENT,
   SFP_IS_PRESENT_ALL,
@@ -118,88 +102,17 @@ static ssize_t set_w_port_data(struct device *dev, struct device_attribute *dev_
 {
   long data;
   int error;
-  long port_t = 0;
-  u8 reg_t = 0x00;
 
   error = kstrtol(buf, 10, &data);
   if(error)
     return error;
 
-  port_t = data;
-
-  if(port_t > 0 && port_t < 9) 
-  {		 /* SFP Port 1-8 */
-    reg_t = QSFP_SFP_SEL_I2C_MUX;
-	} 
-  else if (port_t > 8 && port_t < 17) 
-  {  /* SFP Port 9-16 */
-    reg_t = QSFP_SFP_SEL_I2C_MUX;
-  } 
-  else if (port_t > 16 && port_t < 25) 
-  { /* SFP Port 17-24 */
-    reg_t = QSFP_SFP_SEL_I2C_MUX;
-  } 
-  else if (port_t > 24 && port_t < 33) 
-  { /* SFP Port 25-32 */
-    reg_t = QSFP_SFP_SEL_I2C_MUX;
-  }
-  else if (port_t > 32 && port_t < 37) 
-  { /* SFP Port 33-36 */
-    reg_t = QSFP_SFP_SEL_I2C_MUX;
-  } 
-  else if (port_t > 36 && port_t < 45) 
-  { /* SFP Port 37-44 */
-    reg_t = QSFP_SFP_SEL_I2C_MUX;
-  } 
-  else if (port_t > 44 && port_t < 49) 
-  { /* SFP Port 45-48 */
-    reg_t = QSFP_SFP_SEL_I2C_MUX;
-  } 
-  else if (port_t > 48 && port_t < 55) 
-  { /* QSFP Port 49-54 */
-    reg_t = QSFP_SFP_SEL_I2C_MUX;
-  } 
-  else 
-  {
-
-		/* Disable QSFP SFP channel */
-    if (i2c_cpld_write(I2C_BUS_2, MASTER_CPLD, QSFP_SFP_SEL_I2C_MUX, DEFAULT_DISABLE) < 0) 
-    {
-      return -EIO;
-    }
-
-    goto exit;
-  }
-
-	/* Disable QSFP SFP channel */
-  if (i2c_cpld_write(I2C_BUS_2, MASTER_CPLD, QSFP_SFP_SEL_I2C_MUX, DEFAULT_DISABLE) < 0) 
-  {
-    return -EIO;
-  }
-
-	/* Select SFP or QSFP port channel */
-  if (i2c_cpld_write(I2C_BUS_2, MASTER_CPLD, reg_t, cpld_to_port_table[port_t]) < 0) 
-  {
-    return -EIO;
-  }
-
-exit:
-  sfp_port_data = data;	
+  sfp_port_data = data;
   return count;
 }
 
 static ssize_t for_r_port_data(struct device *dev, struct device_attribute *dev_attr, char *buf)
 {
-
-	if (sfp_port_data == DEFAULT_DISABLE) 
-  {
-		/* Disable QSFP SFP channel */
-    if (i2c_cpld_write(I2C_BUS_2, MASTER_CPLD, QSFP_SFP_SEL_I2C_MUX, DEFAULT_DISABLE) < 0) 
-    {
-      return -EIO;
-    }
-  }
- 
   return sprintf(buf, "%d\n", (int)sfp_port_data);
 }
 
@@ -458,131 +371,7 @@ static ssize_t for_status(struct device *dev, struct device_attribute *dev_attr,
 	}
 }
 
-static ssize_t for_eeprom(struct device *dev, struct device_attribute *dev_attr,char *buf)
-{
-  struct ag5648_sfp_data *data = ag5648_sfp_update_device(dev);
-  if (!data->valid) 
-  {
-    return 0;
-  }
-  memcpy(buf, data->eeprom, sizeof(data->eeprom));
-  return sizeof(data->eeprom);
-}
-
-static int ag5648_sfp_read_block(struct i2c_client *client, u8 command, u8 *data, int data_len)
-{
-  int result = i2c_smbus_read_i2c_block_data(client, command, data_len,data);
-  if (unlikely(result < 0))
-    goto abort;
-  if (unlikely(result != data_len)) 
-  {
-    result = -EIO;
-    goto abort;
-  }
-  result = 0;
-  abort:
-    return result;
-}
-
-static struct ag5648_sfp_data *ag5648_sfp_update_device(struct device *dev)
-{
-  struct i2c_client *client = to_i2c_client(dev);
-  struct ag5648_sfp_data *data = i2c_get_clientdata(client);
-  long port_t = 0;
-	u8 reg_t = 0x00;
-
-  port_t = sfp_port_data;
-
-	memset(data->eeprom, 0, sizeof(data->eeprom));
-	
-  if (port_t > 0 && port_t < 9) 
-  {		 /* SFP Port 1-8 */
-    reg_t = QSFP_SFP_SEL_I2C_MUX;
-	} 
-  else if (port_t > 8 && port_t < 17) 
-  {  /* SFP Port 9-16 */
-    reg_t = QSFP_SFP_SEL_I2C_MUX;
-	} 
-  else if (port_t > 16 && port_t < 25) 
-  { /* SFP Port 17-24 */
-    reg_t = QSFP_SFP_SEL_I2C_MUX;
-	} 
-  else if (port_t > 24 && port_t < 33) 
-  { /* SFP Port 25-32 */
-    reg_t = QSFP_SFP_SEL_I2C_MUX;
-	} 
-  else if (port_t > 32 && port_t < 37) 
-  { /* SFP Port 33-36 */
-    reg_t = QSFP_SFP_SEL_I2C_MUX;
-	} 
-  else if (port_t > 36 && port_t < 45) 
-  { /* SFP Port 37-44 */
-		reg_t = QSFP_SFP_SEL_I2C_MUX;
-	} 
-  else if (port_t > 44 && port_t < 49) 
-  { /* SFP Port 45-48 */
-		reg_t = QSFP_SFP_SEL_I2C_MUX;
-	} 
-  else if (port_t > 48 && port_t < 55) 
-  { /* QSFP Port 49-54 */
-		reg_t = QSFP_SFP_SEL_I2C_MUX;
-	} 
-  else 
-  {
-		memset(data->eeprom, 0, sizeof(data->eeprom));
-
-		/* Disable QSFP SFP channel */
-    if (i2c_cpld_write(I2C_BUS_2, MASTER_CPLD, QSFP_SFP_SEL_I2C_MUX,DEFAULT_DISABLE) < 0) 
-    {
-      goto exit;
-    }
-
-		goto exit;
-	}
-
-	/* Disable QSFP SFP channel */
-	if (i2c_cpld_write(I2C_BUS_2, MASTER_CPLD, QSFP_SFP_SEL_I2C_MUX,DEFAULT_DISABLE) < 0) {
-    memset(data->eeprom, 0, sizeof(data->eeprom));
-		goto exit;
-	}
-
-	/* Select SFP or QSFP port channel */
-  if (i2c_cpld_write(I2C_BUS_2, MASTER_CPLD, reg_t,cpld_to_port_table[port_t]) < 0) 
-  {
-		memset(data->eeprom, 0, sizeof(data->eeprom));
-		goto exit;
-	}
-    mutex_lock(&data->update_lock);
-
-    if (time_after(jiffies, data->last_updated) || !data->valid) 
-    {
-		  int status = -1;
-		  int i = 0;
-		  data->valid = 0;
-      memset(data->eeprom, 0, sizeof(data->eeprom));
-
-      for (i = 0; i < sizeof(data->eeprom)/I2C_SMBUS_BLOCK_MAX; i++) 
-      {
-        status = ag5648_sfp_read_block(client,i * I2C_SMBUS_BLOCK_MAX,data->eeprom + (i * I2C_SMBUS_BLOCK_MAX),I2C_SMBUS_BLOCK_MAX);
-        if (status < 0) 
-        {
-          printk(KERN_INFO "status = %d\n", status);
-          dev_dbg(&client->dev,"unable to read eeprom from port(%d)\n", data->port);
-
-          goto exit;
-        }
-      }
-      data->last_updated = jiffies;
-      data->valid = 1;
-	  }
-
-  exit:
-    mutex_unlock(&data->update_lock);
-    return data;
-}
-
 /* sysfs attributes for hwmon */
-static SENSOR_DEVICE_ATTR(sfp_eeprom,S_IRUGO, for_eeprom, NULL,SFP_EEPROM);
 static SENSOR_DEVICE_ATTR(sfp_select_port, S_IWUSR | S_IRUGO, for_r_port_data, set_w_port_data, SFP_SELECT_PORT);
 static SENSOR_DEVICE_ATTR(sfp_is_present, S_IRUGO, for_status, NULL, SFP_IS_PRESENT);
 static SENSOR_DEVICE_ATTR(sfp_is_present_all, S_IRUGO, for_status, NULL, SFP_IS_PRESENT_ALL);
@@ -590,7 +379,6 @@ static SENSOR_DEVICE_ATTR(sfp_lp_mode,S_IWUSR | S_IRUGO, for_status, set_w_lp_mo
 static SENSOR_DEVICE_ATTR(sfp_reset, S_IWUSR | S_IRUGO, for_status, set_w_reset_data, SFP_RESET);
 
 static struct attribute *ag5648_sfp_attributes[] = {
-	&sensor_dev_attr_sfp_eeprom.dev_attr.attr,
 	&sensor_dev_attr_sfp_select_port.dev_attr.attr,
 	&sensor_dev_attr_sfp_is_present.dev_attr.attr,
 	&sensor_dev_attr_sfp_is_present_all.dev_attr.attr,
@@ -603,24 +391,40 @@ static const struct attribute_group ag5648_sfp_group = {
 	.attrs = ag5648_sfp_attributes,
 };
 
+static int ag5648_sfp_mux_select(struct i2c_mux_core *muxc, u32 chan)
+{
+	return i2c_cpld_write(I2C_BUS_2, MASTER_CPLD, QSFP_SFP_SEL_I2C_MUX, chan);
+}
+
+static int ag5648_sfp_mux_deselect(struct i2c_mux_core *muxc, u32 chan)
+{
+	return i2c_cpld_write(I2C_BUS_2, MASTER_CPLD, QSFP_SFP_SEL_I2C_MUX, DEFAULT_DISABLE);
+}
+
 static int ag5648_sfp_probe(struct i2c_client *client, const struct i2c_device_id *id)
 {
+	struct i2c_adapter *adap = client->adapter;
 	struct ag5648_sfp_data *data;
-	int status;
+	struct i2c_mux_core *muxc;
+	int status, i;
 	
 	if (!i2c_check_functionality(client->adapter,I2C_FUNC_SMBUS_I2C_BLOCK)) 
   {
+	  	dev_err(&client->dev, "failed func\n");
 		status = -EIO;
 		goto exit;
 	}
 	
-	data = kzalloc(sizeof(struct ag5648_sfp_data), GFP_KERNEL);
-	if (!data) {
+	muxc = i2c_mux_alloc(adap, &client->dev, 54,
+			     sizeof(*data), I2C_MUX_LOCKED, ag5648_sfp_mux_select,
+			     ag5648_sfp_mux_deselect);
+	if (!muxc) {
 		status = -ENOMEM;
 		goto exit;
 	}
 
-	mutex_init(&data->update_lock);
+	data = i2c_mux_priv(muxc);
+
 	data->port = id->driver_data;
 	i2c_set_clientdata(client, data);
 
@@ -628,7 +432,7 @@ static int ag5648_sfp_probe(struct i2c_client *client, const struct i2c_device_i
 
 	status = sysfs_create_group(&client->dev.kobj, &ag5648_sfp_group);
   if (status)
-    goto exit_sysfs_create_group;
+    goto exit;
 
 	data->hwmon_dev = hwmon_device_register(&client->dev);
   if (IS_ERR(data->hwmon_dev)) 
@@ -637,14 +441,21 @@ static int ag5648_sfp_probe(struct i2c_client *client, const struct i2c_device_i
     goto exit_hwmon_device_register;
   }
 
+  	for (i = 1; i <= 54; i++) {
+		status = i2c_mux_add_adapter(muxc, 9 + i, i);
+		if (status)
+			goto exit_del_mux_adapters;
+	}
+
 	dev_info(&client->dev, "%s: sfp '%s'\n", dev_name(data->hwmon_dev),client->name);
 
+
 	return 0;
+  exit_del_mux_adapters:
+	  i2c_mux_del_adapters(muxc);
 
   exit_hwmon_device_register:
 	  sysfs_remove_group(&client->dev.kobj, &ag5648_sfp_group);
-  exit_sysfs_create_group:
-	  kfree(data);
   exit:
 	  return status;
 }
diff --git a/packages/platforms/delta/x86-64/ag5648/onlp/builds/x86_64_delta_ag5648/module/src/platform_lib.h b/packages/platforms/delta/x86-64/ag5648/onlp/builds/x86_64_delta_ag5648/module/src/platform_lib.h
index f686ae736fde..97abc270cdfc 100755
--- a/packages/platforms/delta/x86-64/ag5648/onlp/builds/x86_64_delta_ag5648/module/src/platform_lib.h
+++ b/packages/platforms/delta/x86-64/ag5648/onlp/builds/x86_64_delta_ag5648/module/src/platform_lib.h
@@ -71,12 +71,12 @@
 #define FAN4_REAR      PREFIX_PATH "/5-004d/fan4_input"
 #define IDPROM_PATH    "/sys/devices/pci0000:00/0000:00:13.0/i2c-1/i2c-2/2-0053/eeprom"
 
-#define SFP_SELECT_PORT_PATH    PREFIX_PATH "/4-0050/sfp_select_port"
-#define SFP_IS_PRESENT_PATH     PREFIX_PATH "/4-0050/sfp_is_present"
-#define SFP_IS_PRESENT_ALL_PATH PREFIX_PATH "/4-0050/sfp_is_present_all"
-#define SFP_EEPROM_PATH         PREFIX_PATH "/4-0050/sfp_eeprom"
-#define SFP_RESET_PATH          PREFIX_PATH "/4-0050/sfp_reset"
-#define SFP_LP_MODE_PATH        PREFIX_PATH "/4-0050/sfp_lp_mode"
+#define SFP_SELECT_PORT_PATH    PREFIX_PATH "/4-0031/sfp_select_port"
+#define SFP_IS_PRESENT_PATH     PREFIX_PATH "/4-0031/sfp_is_present"
+#define SFP_IS_PRESENT_ALL_PATH PREFIX_PATH "/4-0031/sfp_is_present_all"
+#define SFP_EEPROM_PATH         PREFIX_PATH "/4-0031/sfp_eeprom"
+#define SFP_RESET_PATH          PREFIX_PATH "/4-0031/sfp_reset"
+#define SFP_LP_MODE_PATH        PREFIX_PATH "/4-0031/sfp_lp_mode"
 
 /* BUS define */
 #define I2C_BUS_0             (0)
diff --git a/packages/platforms/delta/x86-64/ag5648/onlp/builds/x86_64_delta_ag5648/module/src/sfpi.c b/packages/platforms/delta/x86-64/ag5648/onlp/builds/x86_64_delta_ag5648/module/src/sfpi.c
index beefbfcf4fc3..0aa6f9feaf4e 100755
--- a/packages/platforms/delta/x86-64/ag5648/onlp/builds/x86_64_delta_ag5648/module/src/sfpi.c
+++ b/packages/platforms/delta/x86-64/ag5648/onlp/builds/x86_64_delta_ag5648/module/src/sfpi.c
@@ -31,6 +31,7 @@
 #include <sys/ioctl.h>
 #include <math.h>
 #include <onlplib/i2c.h>
+#include <onlplib/file.h>
 #include "platform_lib.h"
 /******************* Utility Function *****************************************/
 
@@ -167,13 +168,15 @@ onlp_sfpi_presence_bitmap_get(onlp_sfp_bitmap_t* dst)
     return ONLP_STATUS_OK;
 }
 
+#define PORT_EEPROM_FORMAT "/sys/bus/i2c/devices/%d-0050/eeprom"
+
 int
 onlp_sfpi_eeprom_read(int port, uint8_t data[256])
 {
-    char port_data[3];
+    int bus = (port - 1) + 10;
     int sfp_respond_reg;
     int sfp_respond_val;
-
+    int size = 0;
 
     /* Get respond register if port have it */
     sfp_respond_reg = ag5648_get_respond_reg(port);
@@ -182,20 +185,21 @@ onlp_sfpi_eeprom_read(int port, uint8_t data[256])
     sfp_respond_val = ag5648_get_respond_val(port);
     dni_lock_cpld_write_attribute(MASTER_CPLD_PATH, sfp_respond_reg, sfp_respond_val);
 
-
-    /* Select port */
-    snprintf(port_data, sizeof(port_data), "%d", port);
-    dni_i2c_lock_write_attribute(NULL, port_data, SFP_SELECT_PORT_PATH);
-
     memset(data, 0 ,256);
 
     /* Read eeprom information into data[] */
-    if (dni_i2c_read_attribute_binary(SFP_EEPROM_PATH, (char *)data, 256, 256)
-        != 0)
+    if (onlp_file_read(data, 256, &size, PORT_EEPROM_FORMAT, bus) != ONLP_STATUS_OK)
     {
         AIM_LOG_INFO("Unable to read eeprom from port(%d)\r\n", port);
         return ONLP_STATUS_E_INTERNAL;
     }
+
+    if (size != 256)
+    {
+        AIM_LOG_ERROR("Unable to read the port %d eeprom, unexpected size %i\r\n", port, size);
+        return ONLP_STATUS_E_INTERNAL;
+    }
+
     return ONLP_STATUS_OK;
 }
 
-- 
2.54.0

